core/node/core.go

```diff
@@ -97,11 +102,49 @@ func OnlineExchange(provide bool) interface{} {
 				return exch.Close()
 			},
 		})
+		lc.Append(fx.Hook{
+			OnStart: func(ctx context.Context) error {
+				go func() {
+					for {
+						// We do the check every X seconds. This time is also
+						// as much as we are willing to wait for the response
+						// on all CIDs requested. This means we only do *one*
+						// test at a time.
+						testContext, _ := context.WithTimeout(ctx, time.Second*5) // fixme: parameter, make this a config so we can adjust on the fly
+
+						checkBitswapResponse(testContext, host.ID())
+
+						select {
+						case <-ctx.Done():
```

Comment:
So... this is one of the quirks of fx. This context will (well, may...) get canceled once we're finished starting.

Either call `helpers.LifecycleCtx(lc)`, or, e.g., close a channel from an `OnStop` hook.

Reply:
Ouch! Thanks for flagging this, I would have never expected this behavior:

https://github.com/uber-go/fx/blob/4adc2bf73f2cdcd423bf22725985177670a91d4d/app.go#L738-L742

Reply:
Yeah, I'm pretty sure we have a few bugs due to this in lotus.